pkg/expansion/expand.go

```diff
@@ -0,0 +1,224 @@
+package expansion
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+
+	mutationsunversioned "github.com/open-policy-agent/gatekeeper/apis/mutations/unversioned"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutators/assign"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutators/assignmeta"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutators/modifyset"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var MutatorTypes = map[string]bool{"Assign": true, "AssignMetadata": true, "ModifySet": true}
+
+func sourcePath(source string) []string {
+	return strings.Split(source, ".")
+}
+
+func prettyResource(v interface{}) string {
+	b, err := json.MarshalIndent(v, "", "  ")
+	if err == nil {
+		return string(b) + "\n"
+	}
+	return ""
+}
+
+func ExpandResources(resources []*unstructured.Unstructured) ([]*unstructured.Unstructured, error) {
+	generators, unstructMutators, unstructTemplates := sortResources(resources)
+
+	mutators, err := convertMutators(unstructMutators)
+	if err != nil {
+		return nil, fmt.Errorf("error converting mutators: %s", err)
+	}
+	templates, err := convertTemplateExpansions(unstructTemplates)
+	if err != nil {
+		return nil, fmt.Errorf("error converting template expansions: %s", err)
+	}
+
+	cache, err := NewExpansionCache(mutators, templates)
+	if err != nil {
+		return nil, fmt.Errorf("error creating Cache: %s", err)
+	}
+
+	var resultants []*unstructured.Unstructured
+	for _, gen := range generators {
+		temps := cache.TemplatesForGVK(gen.GroupVersionKind())
+		for i := 0; i < len(temps); i++ {
+			t := temps[i]
+			muts := cache.MutatorsForGVK(t.Spec.GeneratedGVK)
+			result, err := ExpandGenerator(gen, &t, muts)
+			if err != nil {
+				return nil, fmt.Errorf("error expanding generator: %s", err)
+			}
+			resultants = append(resultants, result)
+		}
+	}
+
+	return resultants, nil
+}
+
+func ExpandGenerator(generator *unstructured.Unstructured, template *mutationsunversioned.TemplateExpansion, mutators []types.Mutator) (*unstructured.Unstructured, error) {
+	srcPath := template.Spec.TemplateSource
+	if srcPath == "" {
+		return nil, fmt.Errorf("cannot expand generator for template with no source")
+	}
+	resultantGVK := template.Spec.GeneratedGVK
+	emptyGVK := schema.GroupVersionKind{}
+	if resultantGVK == emptyGVK {
+		return nil, fmt.Errorf("cannot expand generator for template with empty generatedGVK")
+	}
+
+	src, ok, err := unstructured.NestedMap(generator.Object, sourcePath(srcPath)...)
```

Comment:
This will break if we attempt to access a template source under a list index. That might be fine, but we should document/test for that limitation.